### PR TITLE
Fixes #18591 - powerStatus in redux

### PR DESCRIPTION
**Pull request by matanwerbner**
also some refactoring on statistics redux implementation and loader

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/hosts/PowerStatus.css [hunk 1]
```diff
@@ -1,9 +1,11 @@
 .host-power-status.on {
-  color: #3f9c35;
+    color: #3f9c35;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/hosts/PowerStatus.css]: Please use 2 spaces vs 4

File: webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.js [hunk 2]
```diff
@@ -26,7 +27,7 @@ class StatisticsChartsList extends React.Component {
   render() {
     const noDataMsg = __('No data available').toString();
     const tip = __('Expand the chart').toString();
-    const charts = this.props.charts.map(chart => {
+    const charts = map(this.props.charts, chart => {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.js]: or what is the difference between the two "maps"?

File: webpack/assets/javascripts/react_app/components/hosts/powerStatus/PowerStatus.css [hunk 3]
```diff
@@ -1,9 +1,11 @@
 .host-power-status.on {
-  color: #3f9c35;
+    color: #3f9c35;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/hosts/powerStatus/PowerStatus.css]: please keep 2 spaces for css.

File: webpack/assets/javascripts/react_app/stores/HostsStore.js [hunk 4]
```diff
@@ -1,62 +0,0 @@
...
-  // workaround for max listeners
-  maxListers(selector = 'meta[name=itemsPerPage]') {
-    const element = document.querySelector(selector);
-
-    if (element && element.content) {
-      // * 2 as we have both success and failure listeners
-      return (element.content * 2);
-    }
-
-    return 40;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/stores/HostsStore.js:28]: could you also remove the meta attribute that was introduced as part of the original PR - see https://github.com/theforeman/foreman/commit/901fe896ae50ed3c75ee0a86bc78463f9973dc76#diff-7eb571b855131f643240b7b93ddfd44dR3

**Review by ohadlevy — CHANGES_REQUESTED** [review]

**Comment by ohadlevy** [comment]
@gailsteiger can you review please? thanks

File: webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.js (hunk 2 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.js]: when the state was an array, array.map is used.
after this commit the state will be an object key/value, so lodash map is used

Comment (gailsteiger) [reply webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.js]: In this case, I think it is better to use the `_.map(...)` syntax to make the lodash dependency more explicit.

**Comment by gailsteiger** [comment]
Is 'webpack/assets/javascripts/react_app/actions/HostsActions.js' used?

Same question for  'receivedHostsPowerState' and 'hostsRequestError' in 'webpack/assets/javascripts/react_app/actions/ServerActions.js'

**Comment by matanwerbner** [comment]
@gailsteiger not sure about lodash, but i've done what you suggested.
i think it's slightly better to use map() instead of _.map(), since the first does not couple you to lodash, and you are free to import or implement map or other such methods.

**Comment by gailsteiger** [comment]
@matanwerbner Thanks for making the change to use the explicit syntax with the map function. I understand your point. I'm less concerned about coupling at this point.

**Comment by ohadlevy** [comment]
one note, I see 
```warning "sass-loader@6.0.2" has incorrect peer dependency "webpack@^2.0.0".```
does it mean we need to restrict the version of sass-loader to work on our current version of webpack?

**Comment by ohadlevy** [comment]
@matanwerbner can you please:
1. rebase
2. update storybook as well?
thanks.

File: webpack/assets/javascripts/react_app/components/common/Loader.js [hunk 5]
```diff
@@ -3,32 +3,40 @@ import {STATUS} from '../../constants';
...
-const Loader = ({ status, children, spinnerSize = 'lg' }) => {
-let content;
+const wrapWithLoaderRoot = content => (
+  <div className="loader-root">
+    {content}
+  </div>
+);
+
+export const simpleLoader = (spinnerSize) => wrapWithLoaderRoot(
+  <div className={`spinner spinner-${spinnerSize}`}></div>
```
Comment (gailsteiger) [inline webpack/assets/javascripts/react_app/components/common/Loader.js]: @matanwerbner I submitted a [PR](https://github.com/theforeman/foreman/pull/4343) to create a Spinner component.

It's very simple. If it is merged, you can replace line 13 with
```
<Spinner size={spinnerSize} />
```

**Comment by matanwerbner** [comment]
@ohadlevy done. please note that storybook is failing in devleop because of incompatibility with scss, i have put a seperate small PR that fixes it. 
about sass-loader, i think it's a different PR, no?

**Comment by dLobatog** [comment]
@matanwerbner Mind to rebase and push again to make tests pass? The latest changes in develop contain your fix for sass in the storybook. Thanks!

**Comment by matanwerbner** [comment]
@dLobatog rebased and pushed, check again?

File: config/webpack.config.js [hunk 6]
```diff
@@ -96,7 +96,9 @@ if (production) {
   config.devServer = {
     host: process.env.BIND || '127.0.0.1',
     port: devServerPort,
-    headers: { 'Access-Control-Allow-Origin': '*' }
+    headers: { 'Access-Control-Allow-Origin': '*' },
+    hot: true,
+    inline: true
```
Comment (ohadlevy) [inline config/webpack.config.js]: why are those required? should they be in the same patch?

File: webpack/assets/javascripts/react_app/redux/actions/hosts/hosts.test.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+import * as actions from './index';
+import immutable from 'seamless-immutable';
+import { requestData, onFailureActions } from './hosts.fixtures';
+const mockStore = configureMockStore([thunk]);
+
+describe('hosts actions', () => {
+  it(
+    'creates HOST_POWER_STATUS_REQUEST and fails when nock is not applied',
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/hosts/hosts.test.js]: what is nock?

File: webpack/assets/javascripts/react_app/redux/reducers/hosts/index.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+export default (state = initialState, action) => {
+  const { payload } = action;
+
+  switch (action.type) {
+    case HOST_POWER_STATUS_REQUEST:
+      return state.setIn(
+        ['powerStatus', payload.id],
+        payload
+      );
+    case HOST_POWER_STATUS_SUCCESS:
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/reducers/hosts/index.js]: since this one and the one above are the same, DRY?

File: webpack/stories/index.js [hunk 9]
```diff
@@ -77,41 +78,29 @@ storiesOf('Charts', module)
...
+  .add('Loading', () => simpleLoader('xs'))
   .add('ON', () => (
-    <PowerStatus
-      state="on"
-      title="on"
-      loadingStatus="RESOLVED"
-      statusText="On"
-    />
+    <span
+        className={'fa fa-power-off host-power-status on'}
```
Comment (ohadlevy) [inline webpack/stories/index.js]: hmmm.. this is not using the components at all?? the whole point was to reuse the components not just to show css?

File: webpack/assets/javascripts/react_app/redux/actions/hosts/hosts.test.js (hunk 7 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/redux/actions/hosts/hosts.test.js]: HTTP mocking lib, since this action triggers a AJAX request, nock can intercept and create a response. if it is not used, error will be returned, and this is the case i'm testing.

File: webpack/assets/javascripts/react_app/redux/reducers/hosts/index.js (hunk 8 above)
Comment (gailsteiger) [reply webpack/assets/javascripts/react_app/redux/reducers/hosts/index.js]: I understand Ohad's comment but I question whether or not we what to allow breakthrough case statements. Perhaps in this case wetter may be better?

**Comment by matanwerbner** [comment]
Guys, anything else? @ohadlevy @gailsteiger

**Comment by gailsteiger** [comment]
It's OK with me.

**Comment by ohadlevy** [comment]
@matanwerbner can you please rebase the pr, I think it will fix the test failure.

**Comment by matanwerbner** [comment]
@ohadlevy  anything else? can we merge?

**Comment by ohadlevy** [comment]
@matanwerbner when running this on my instance i get the following failure - am i missing something?
```js
Uncaught TypeError: Cannot read property 'markup' of undefined
    at Object.mount (MountingService.js:32)
    at <anonymous>:3:20
    at executeScriptTags (turbolinks.self.js?body=1:227)
    at changePage (turbolinks.self.js?body=1:197)
    at XMLHttpRequest.xhr.onload (turbolinks.self.js?body=1:107)
```

**Review by ohadlevy — APPROVED** [review]
@matanwerbner I take it back, some local issues, work well, thanks!
